### PR TITLE
New and improved takeoff controller

### PR DESCRIPTION
**Pull request by aftersomemath**

### CODE REVIEW

File: param/low_level_motion_0.05.yaml [hunk 1]
```diff
@@ -58,6 +58,9 @@ update_frequency: 60.0
 startup_timeout: 10.0
 update_timeout: 0.2
 
+# Time to delay after arming
+post_arm_delay: 0.2
```
Comment (asaba96) [inline param/low_level_motion_0.05.yaml]: units?

File: param/low_level_motion_1.1.yaml [hunk 2]
```diff
@@ -51,12 +52,21 @@ yaw_max           : 0.5
...
+level_flight_required_hysteresis: 0.10
+
 update_frequency: 60.0
 startup_timeout: 10.0
 update_timeout: 0.2
 
-# Rate at which to ramp the throttle during takeoff in percent a second
-takeoff_throttle_ramp_rate: 0.1
+# Time to delay after arming
+post_arm_delay: 0.2
```
Comment (asaba96) [inline param/low_level_motion_1.1.yaml]: same as above

File: param/low_level_motion_1.1.yaml [hunk 3]
```diff
@@ -51,12 +52,21 @@ yaw_max           : 0.5
...
+
+# Time to take to ramp the throttle to hover throttle
+takeoff_throttle_ramp_duration: 1.0
```
Comment (asaba96) [inline param/low_level_motion_1.1.yaml]: units?

File: param/low_level_motion_crazyflie.yaml [hunk 4]
```diff
@@ -16,23 +16,24 @@ thrust_model:
     voltage_polynomial: [1.0]
     throttle_b: 0.34352
     throttle_c: -0.24205
+    expected_hover_throttle: 0.57
```
Comment (asaba96) [inline param/low_level_motion_crazyflie.yaml]: how was this calculated?

File: param/low_level_motion_crazyflie.yaml [hunk 5]
```diff
@@ -41,24 +42,33 @@ throttle_max      : 1.0
...
+roll_max          : 0.5
+roll_min          : -0.5
+roll_max_rate     : 10.0
 
 yaw_max           : 0.5
 yaw_min           : -0.5
 yaw_max_rate      : 10.0
 
+# Height below which the drone is required to be level
+level_flight_required_height: 0.15
```
Comment (asaba96) [inline param/low_level_motion_crazyflie.yaml]: Units?

File: param/low_level_motion_crazyflie.yaml [hunk 6]
```diff
@@ -41,24 +42,33 @@ throttle_max      : 1.0
...
+# Height above the level flight height at which to allow the drone
+# to begin tilting again
+level_flight_required_hysteresis: 0.10
```
Comment (asaba96) [inline param/low_level_motion_crazyflie.yaml]: again, units?

File: param/low_level_motion_crazyflie.yaml [hunk 7]
```diff
@@ -41,24 +42,33 @@ throttle_max      : 1.0
...
+
 update_frequency: 60.0
 startup_timeout: 10.0
 update_timeout: 0.2
 
-# Rate at which to ramp the throttle during takeoff in percent a second
-takeoff_throttle_ramp_rate: 0.1
+# Time to delay after arming
+post_arm_delay: 0.2
```
Comment (asaba96) [inline param/low_level_motion_crazyflie.yaml]: same as above.

File: param/low_level_motion_crazyflie.yaml [hunk 8]
```diff
@@ -41,24 +42,33 @@ throttle_max      : 1.0
...
+
+# Time to take to ramp the throttle to hover throttle
+takeoff_throttle_ramp_duration: 1.0
```
Comment (asaba96) [inline param/low_level_motion_crazyflie.yaml]: same as above.

File: param/low_level_motion_sim.yaml [hunk 9]
```diff
@@ -51,12 +52,21 @@ yaw_max           : 3.5
 yaw_min           : -3.5
 yaw_max_rate      : 1000.0
 
+# Height below which the drone is required to be level
+level_flight_required_height: 0.35
```
Comment (asaba96) [inline param/low_level_motion_sim.yaml]: same as above.

File: param/low_level_motion_sim.yaml [hunk 10]
```diff
@@ -51,12 +52,21 @@ yaw_max           : 3.5
...
+# Height above the level flight height at which to allow the drone
+# to begin tilting again
+level_flight_required_hysteresis: 0.10
```
Comment (asaba96) [inline param/low_level_motion_sim.yaml]: same as above.

File: param/low_level_motion_sim.yaml [hunk 11]
```diff
@@ -51,12 +52,21 @@ yaw_max           : 3.5
...
+
 update_frequency: 60.0
 startup_timeout: 10.0
 update_timeout: 2.0
 
-# Rate at which to ramp the throttle during takeoff in (percent/100) a second
-takeoff_throttle_ramp_rate: 0.5
+# Time to delay after arming
+post_arm_delay: 0.2
```
Comment (asaba96) [inline param/low_level_motion_sim.yaml]: same as above.

File: param/low_level_motion_sim.yaml [hunk 12]
```diff
@@ -51,12 +52,21 @@ yaw_max           : 3.5
...
+
+# Time to take to ramp the throttle to hover throttle
+takeoff_throttle_ramp_duration: 1.0
```
Comment (asaba96) [inline param/low_level_motion_sim.yaml]: same as above.

File: param/tasks.yaml [hunk 13]
```diff
@@ -9,7 +9,11 @@ takeoff_velocity        : 0.3
 takeoff_acceleration    : 0.3
 # Height at which to consider takeoff finished in meters
 # Should be higher than min_maneuver_height
-takeoff_complete_height: 0.5
+takeoff_complete_height: 0.8
+# Deviance from takeoff height from which to enter stabilize mode
+takeoff_complete_height_tolerance: 0.15
+# Length of time to stabilize after takeoff
+takeoff_stabilize_delay: 1.5
```
Comment (asaba96) [inline param/tasks.yaml]: Units please for these 3 params

File: param/tasks.yaml [hunk 14]
```diff
@@ -59,9 +63,16 @@ drone_landing_gear_width: .775
 
 # HEIGHT HOLDER
 # p-term for height holder helper for Z height
-p_term_height_hold_z: .90
+p_term_height_hold_z: 2.0
 # maximum z error for heigth holder
 max_z_error: 1.0
+# distance from setpoint that will allow deadzone mode to begin
+deadzone_height_hold_z: 0.05
```
Comment (asaba96) [inline param/tasks.yaml]: Units. Also, why did you choose these names? what does `deadzone` mean?

File: src/TakeoffController.cpp [hunk 15]
```diff
@@ -160,25 +162,51 @@ bool TakeoffController::update(const ros::Time& time,
...
+                                                                 update_timeout_);
+            if (!success) {
+                ROS_ERROR("Failed to get current transform to calibrate thrust model");
+                return false;
+            }
+
+            geometry_msgs::PointStamped col_point;
+            tf2::doTransform(col_point, col_point, transform);
+
+            double hover_throttle = thrust_model_.throttleFromAccel(9.8, voltage, col_point.point.z);
```
Comment (asaba96) [inline src/TakeoffController.cpp]: is the 9.8 the acceleration due to gravity? is this how we want to do this? if so, wouldn't it be better to provide more precision (i.e. 9.81 m/s^2)

File: src/TakeoffController.cpp [hunk 16]
```diff
@@ -160,25 +162,51 @@ bool TakeoffController::update(const ros::Time& time,
...
+
+            throttle_ = ((time-ramp_start_time_).toSec()/1.0) * hover_throttle;
```
Comment (asaba96) [inline src/TakeoffController.cpp]: comment here please

File: src/TakeoffController.cpp [hunk 17]
```diff
@@ -251,8 +278,9 @@ bool TakeoffController::waitUntilReady()
     }
 
     // This time is just used to calculate any ramping that needs to be done.
-    last_update_time_ = std::max(landing_detected_message_.header.stamp,
-                                 battery_interpolator_.getLastUpdateTime());
+    last_update_time_ = std::max({transform.header.stamp,
+                                  landing_detected_message_.header.stamp,
+                                  battery_interpolator_.getLastUpdateTime()});
```
Comment (asaba96) [inline src/TakeoffController.cpp]: Is this valid syntax? just not sure, given the extra `{}` which was not used before.

File: src/iarc7_motion/iarc_tasks/task_utilities/height_holder.py [hunk 18]
```diff
@@ -18,20 +19,46 @@ def __init__(self, desired_height = None):
...
                 raise ValueError('No height to hold')
             self._delta_z = self._DESIRED_HEIGHT - height
-            return self._K_Z * self._delta_z
+
+            if self._DEBUG:
+                msg = Vector3Stamped()
+                msg.header.stamp = rospy.Time.now()
+                msg.vector.x = self._DESIRED_HEIGHT
+                msg.vector.y = height
+                msg.vector.z = self._delta_z
```
Comment (asaba96) [inline src/iarc7_motion/iarc_tasks/task_utilities/height_holder.py]: the variable names and what you are assigning to them makes my brain hurt :(

File: src/iarc7_motion/iarc_tasks/task_utilities/height_holder.py [hunk 19]
```diff
@@ -18,20 +19,46 @@ def __init__(self, desired_height = None):
...
+                self._debug_pub.publish(msg)
+
+            if abs(self._delta_z) < self._DEADZONE:
```
Comment (asaba96) [inline src/iarc7_motion/iarc_tasks/task_utilities/height_holder.py]: what is `deadzone` ? A comment on what it is would be helpful.

File: src/iarc7_motion/iarc_tasks/velocity_task.py [hunk 20]
```diff
@@ -51,7 +51,12 @@ def __init__(self, task_request):
         self._HORIZ_X_VEL = task_request.x_velocity
         self._HORIZ_Y_VEL = task_request.y_velocity
 
-        self._z_holder = HeightHolder()
+        if task_request.z_position > 0.00001:
```
Comment (asaba96) [inline src/iarc7_motion/iarc_tasks/velocity_task.py]: I'm sure there is a succinct Pythonic-way of doing this, no @amiller27 ?

File: include/iarc7_motion/QuadVelocityController.hpp [hunk 21]
```diff
@@ -118,6 +118,10 @@ class QuadVelocityController
 
     // Max allowed requested thrust in m/s^2
     double max_thrust_;
+
+    const double level_flight_required_height_;
```
Comment (asaba96) [inline include/iarc7_motion/QuadVelocityController.hpp]: please comment on what these represent.

**Comment by asaba96** [comment]
Looks like an improvement. Overall, however, could benefit from better comments and from better variable names, which are not always self-explantory or are very long or ambiguous to what they represent.

File: src/iarc7_motion/iarc_tasks/velocity_task.py (hunk 20 above)
Comment (amiller27) [reply src/iarc7_motion/iarc_tasks/velocity_task.py]: Succinct way of doing what? This looks fine to me

Comment (asaba96) [reply src/iarc7_motion/iarc_tasks/velocity_task.py]: Checking for zero

Comment (amiller27) [reply src/iarc7_motion/iarc_tasks/velocity_task.py]: He's not checking for zero though

File: src/TakeoffController.cpp (hunk 15 above)
Comment (Riteshlogin) [reply src/TakeoffController.cpp]: Actually, we have to figure out what g is equal to at the Georgia Tech venue and at Pitt and make that a param that goes to the lowest digit possible

File: src/iarc7_motion/iarc_tasks/velocity_task.py (hunk 20 above)
Comment (Riteshlogin) [reply src/iarc7_motion/iarc_tasks/velocity_task.py]: But why not check for zero?

File: param/tasks.yaml (hunk 13 above)
Comment (Riteshlogin) [reply param/tasks.yaml]: The response from both of our beloved leads is that we always use mks units, and my response to this is that there should just be two words at the top - "mks units". It's a standard across every single one of our yaml files (as far as I've seen) but it isn't a big deal to have it on top of each file

File: include/iarc7_motion/QuadVelocityController.hpp (hunk 21 above)
Comment (aftersomemath) [reply include/iarc7_motion/QuadVelocityController.hpp]: Done

File: param/low_level_motion_crazyflie.yaml (hunk 4 above)
Comment (aftersomemath) [reply param/low_level_motion_crazyflie.yaml]: It was found by testing.

File: param/tasks.yaml (hunk 14 above)
Comment (aftersomemath) [reply param/tasks.yaml]: Deadzone is an area where the control output is 0. I just came up with the names.

File: src/TakeoffController.cpp (hunk 16 above)
Comment (aftersomemath) [reply src/TakeoffController.cpp]: Fixed

File: src/TakeoffController.cpp (hunk 17 above)
Comment (aftersomemath) [reply src/TakeoffController.cpp]: Aaron actually wrote this, it is an Initialization list. http://en.cppreference.com/w/cpp/utility/initializer_list

File: src/iarc7_motion/iarc_tasks/task_utilities/height_holder.py (hunk 19 above)
Comment (aftersomemath) [reply src/iarc7_motion/iarc_tasks/task_utilities/height_holder.py]: Fixed

File: src/iarc7_motion/iarc_tasks/velocity_task.py (hunk 20 above)
Comment (aftersomemath) [reply src/iarc7_motion/iarc_tasks/velocity_task.py]: Fixed

File: param/tasks.yaml (hunk 13 above)
Comment (aftersomemath) [reply param/tasks.yaml]: We are following REP103 for units so I don't think they need to be specified.

http://www.ros.org/reps/rep-0103.html#units

File: src/iarc7_motion/iarc_tasks/task_utilities/height_holder.py (hunk 18 above)
Comment (aftersomemath) [reply src/iarc7_motion/iarc_tasks/task_utilities/height_holder.py]: Fixed

**Pull request merged**